### PR TITLE
Fix memory allocation functions

### PR DESCRIPTION
**Pull request by rafael2k**
Many fixes for the memory allocation functions. The toolchain is stable when compiling both examples in "examples" folder.

### CODE REVIEW

**Comment by ghaerr** [comment]
Nice, thanks for getting most of the memory allocation code updated in all the tools.

In general, this all looks very good. I have a couple questions/comments that I'll comment on the source directly.

Thank you!

File: assembler/nasmlib.c [hunk 1]
```diff
@@ -95,18 +94,11 @@ void *nasm_realloc (void *q, size_t size)
...
-		{                    /* on fail, previous memory not freed */
-			memcpy(p, q, size);     /* FIXME copies too much!! */
-			nasm_free(q);
-		}
+	p = nasm_malloc(size);
+	if (p)
+	{                    /* on fail, previous memory not freed */
+		memcpy(p, q, size);     /* FIXME copies too much!! */
+		nasm_free(q);
 	}
```
Comment (ghaerr) [inline assembler/nasmlib.c:102]: Do you think we might be better off in NASM to follow the idea you're using in C86? That is just redefine malloc/free/realloc to memalloc/memfree/memrealloc, which directs the function calls directly to mem.c?

In that way, nasmlib.c goes back to exactly how it was when you forked upstream, and all allocations are handled specially in mem.c, which would be ifdef ELKS for host/elks, and then easily replaced by libc code. Currently, when the mem.c code is placed in our C library, the NASM realloc isn't what it started with, but is handled by a slow malloc/free rather than actual realloc, for those systems that have it.

File: compiler/memmgt.c [hunk 2]
```diff
@@ -34,7 +34,7 @@
 /*
  * Memory is allocated in Blocks of 1024 longs, which form a linked list
  */
-#define BLKLEN ((size_t)1024 * sizeof(long))
+#define BLKLEN ((size_t)768 * sizeof(long))
```
Comment (ghaerr) [inline compiler/memmgt.c:37]: Did you find that 4K blocks are too large? Just wondering. In any case, I recommend that you add "// was 1024" so that we know that the original setting was, in case we have to go back to it because of testing.

File: compiler/proto.h [hunk 3]
```diff
@@ -359,10 +359,13 @@ extern void size_type P_ ((TYP *));
...
-#define	malloc(s)		fmemalloc(s)
-#define realloc(p,s)    fmemrealloc((p),(s))
+#define	free(s)			memfree(s)
+#define	malloc(s)		memalloc(s)
+#define realloc(p,s)    memrealloc((p),(s))
 
+void __far *memalloc(unsigned long size);
+void memfree(void __far *ptr);
+void __far *memrealloc(void __far *ptr, unsigned long size);
 #endif
```
Comment (ghaerr) [inline compiler/proto.h:369]: This looks great, but perhaps consider putting the undef's, define's and C declarations all in mem.h and including mem.h into proto.h. That separates out the external API and then only a single file need be changed when the functions are moved into the C library: mem.h and mem.c (both would be deleted in the applications, as they would be replaced by C library code or standard headers).

File: compiler/system.c [hunk 4]
```diff
@@ -53,23 +53,51 @@
 #include "proto.h"
 
 #ifdef __ELKS__
-void *fmemrealloc(void *q, size_t size)
+
+#define MAX_NEAR_ALLOC  64U   /* max size to allocate from near heap */
```
Comment (ghaerr) [inline compiler/system.c:57]: Why is this so small? That's going to cause lots of far heap allocations with small data. I like your reduction to 512U in the mem.c file as a reasonable alternative to 2K, but I'm worried that 64 bytes is going to reintroduce the kernel out of heap errors we saw before when we have a bunch of tiny far allocations.

File: compiler/system.c [hunk 5]
```diff
@@ -53,23 +53,51 @@
...
+	else
+	{
+		new = memalloc(size);
+		memcpy(new, ptr, size);    /* FIXME copies too much!! */
+		memfree(ptr);
+	}
+	return new;
 }
+
 #endif
```
Comment (ghaerr) [inline compiler/system.c:101]: See comment above - how about the idea of moving all this code out of system.c into mem.c? The same exact mem.c that would be used in all the tools. This is the only way we'll know we're ready to remove the memory functions into the C library, because they'd all be exactly the same - in mem.c. This idea also keeps all our special code out of the tool's .c files, providing better separation.

File: cpp/mem.c [hunk 6]
```diff
@@ -26,10 +26,12 @@ void __far *xalloc(unsigned long size)
 	if (size <= MAX_NEAR_ALLOC)
 	{
 		p = malloc((unsigned int)size);
-		if (p == NULL)
-			return (void __far *)p;
+		fp = (void __far *)p;
```
Comment (ghaerr) [inline cpp/mem.c:29]: This cast conversion can be *very* tricky when using near and far pointer conversions, and I don't recommend it. I have seen IA16 and OWC compilers produce the wrong desired output, and the only way to know is by running `objdump86` or `wdis` on the object file to see what code was emitted.

What we're talking about here is that malloc, internally, maintains a near heap, even in large model. So in certain cases, the near NULL pointer "0", gets converted to a far pointer DS:0, which is not 0. Its a far pointer using the near data seg's segment value (DS) and the near NULL pointer, 0, when when combined is NOT 0!

My recommendation to get around this and not by bitten by the C compiler is to not try to convert NULL pointers from near to far. (I can't recall whether this is properly handled by the C library malloc when compiled for large model, which is the point).  In general, when malloc returns NULL in large model, 'p' above will be 0. So, at that point, just return 0, rather than attempt a far conversion. By all rights, the compiler should not generate any code when converting a 'char *' to a 'char far *' in large model, unless for some reason it thinks that malloc returns a near pointer (it does internally). But I'm a bit worried because I had to rewrite malloc for OWC to use `__wcnear` pointers internally - near pointers. (OWC supports the near keyword which allows for near pointers to be manipulated in large model).

Here's the code for ELKS malloc: https://github.com/ghaerr/elks/blob/master/libc/malloc/malloc.c, check out the use of `__wcnear` which is defined to `near` for OWC. Before my conversion to using near pointers, malloc failed miserably in large model. The reason is that the default data segment is managed as a near heap, and has to be, since the normal application only has a single data segment (containing .data, .bss, the heap and stack).

In order to check the the code, I use:
```
$ wdis mem.obj    # for Open Watcom
$ ia16-elf-objdump -D -r -Mi8086 mem.o   # for IA16
```
Sorry for the long comment. I will check further myself what malloc does internally, I am now a bit worried as to what exactly malloc returns (0:0 vs DS:0) in the case of Out of Memory!

File: compiler/memmgt.c (hunk 2 above)
Comment (rafael2k) [reply compiler/memmgt.c:37]: fmemalloc was returning NULL with bigger allocations.

File: compiler/system.c (hunk 4 above)
Comment (rafael2k) [reply compiler/system.c:57]: We need this value parametrized in the mem api. This value was obtained experimentally, decreasing from 1024, the first that worked.

File: cpp/mem.c (hunk 6 above)
Comment (rafael2k) [reply cpp/mem.c:29]: Thanks for this. I'll read carefully. I'll try to fix the near to far conversion issues.

File: compiler/system.c (hunk 5 above)
Comment (rafael2k) [reply compiler/system.c:101]: Yes, but first we need to improve the api for setting the MAX_NEAR_ALLOC, for example.

Comment (ghaerr) [reply compiler/system.c:101]: > first we need to improve the api for setting the MAX_NEAR_ALLOC

Agreed, I didn't quite realize you've not been able to use the same value of MAX_NEAR_ALLOC. But each copy of mem.c could have a different value for MAX_NEAR_ALLOC, with the other portions the same. This would still keep our memory allocation routines out of the tool main source code, for example.

File: compiler/system.c (hunk 4 above)
Comment (ghaerr) [reply compiler/system.c:57]: > We need this value parametrized in the mem api.

Agreed. I was thinking of replacing MAX_NEAR_ALLOC with `unsigned int max_near_alloc = 512` in mem.c, adding the line `extern unsigned int max_near_alloc;` in mem.h, and then adding `max_near_alloc = 64;` in the tool init or main code, something like that. That would allow the same copy of mem.c to be used, but parameterized to a different value at startup.

It doesn't really matter exactly how we do it now, other than my comments are to try to reduce you or I changing it again when the C library code takes over. (Yes, I should have done this a lot earlier in the ELKS C library, but didn't realize just how tight things are and what you've had to go through to get things running!)

All that said, I am quite surprised at having to lower this value to 64 in order for C86 to work! This means that C86 must be allocating lots and lots of allocations > 64, which I find surprising. IIRC only the symbols buffer uses malloc directly in C86, the code generation and other large buffers use a BLKLEN allocation in order to reduce explicit malloc calls  (see my other comment about that).

I will add the capability to see what sizes malloc and fmemalloc are being called with so we can get to the bottom of what is going on with the many (different?) allocations the tools are doing. I've been traveling and haven't had as much time, but its coming!

File: compiler/memmgt.c (hunk 2 above)
Comment (ghaerr) [reply compiler/memmgt.c:37]: Something is definitely fishy here - C86 can't do a 4K (=1024 \* sizeof(long)) far allocation, in all of main memory?! 

As per my other comment, it seems we need to be able to see more of the insides of what allocations are occurring, since I am definitely surprised at not being able to allocate 4K!! All I can think of is that main memory is littered with tons of 65+ byte far allocations from the lowered MAX_NEAR_ALLOC, which you said had to be done.

Overall, it seems managing the toolchain memory might be more complicated than originally thought. I had hoped to be able to keep most small allocations in the near heap, but if we can't since the near heap is super tight, which then causes the far heap to run out of memory when allocating a small 4K, then we might have to move to a more complicated memory management algorithm. In the mean time, I'm glad you've got this working, but it seems we could still easily run out of memory, that things might be fragile, depending on the C source.

What I'm thinking we might have to do would be move to a multi-heap allocation manager. That is, instead of just allocating from the near heap, which sounds like its flat out of space for larger programs (and C86 has lots of strings, so not surprising there), the allocator would allocate larger 8-64k chunks of main memory and use them as "arenas" from which smaller allocations would be allocated from. In this way we can possibly manage tons of small allocations, larger than the near heap in the default data segment, but also not fragment main memory. Lets talk more about all this after I get the malloc visibility enhancements into ELKS.

Thanks for your testing and getting the toolchain back working!

**Comment by ghaerr** [comment]
@rafael2k:

> I will check further myself what malloc does internally, I am now a bit worried as to what exactly malloc returns (0:0 vs DS:0) in the case of Out of Memory!

I must say that while our conversations may sometimes seem long-winded, your heavy use of ELKS to develop this toolchain is definitely bringing some (more) problem ELKS issues to the forefront!

> My recommendation to get around this and not by bitten by the C compiler is to not try to convert NULL pointers from near to far.

In this case, I should take my own advice - the automatic conversion of a near to far pointer in malloc() itself is causing a return of a non-zero DS:0, instead of 0 in large model! Ugh! This means that malloc never returns NULL to any of your routines. Because of that either the memrealloc continues with a bad memcpy to DS:0, or the toolchain tool continues with what it thinks is a non-null pointer, when in fact it is a near-heap, non-zero NULL!! No wonder we're seeing some strange results sometimes.

I plan on fixing all this, but since I brought all this up, I'll show the actual buggy details for your enjoyment.

In libc/malloc/malloc.c, towards the end, cleaned up, we have:
```
void *malloc(size_t size)
{
    register mem __wcnear *ptr = 0; // near pointer used internally. Set to NULL to start
...
     ptr = __search_chunk(sz);  
     if (ptr == 0)                      
     {
        if( ptr == 0) errno = ENOMEM; // 12
        return ptr; // RETURN #1 (see below)
     }
 ...
   return ptr + 1;    // RETURN #2 (see below)                           
}
```
Running `wdis malloc.obj` gives us, at the end: (note SS == DS and 32-bit result is returned in DX:AX)
```
024A  E8 00 00                          call            __search_chunk_
024D  89 C3                             mov             bx,ax // save ptr in BX
024F  85 C0                             test            ax,ax  // AX = PTR
0251  75 0E                             jne             L$32
0253  75 06                             jne             L$31  // second "if (ptr == 0)" optimized out
0255  C7 86 00 00 0C 00                 mov             word ptr _errno[bp],0x000c // errno = ENOMEM
025B                            L$31:
025B  8C D2                             mov             dx,ss // returns far NULL pointer DX:AX as DS:0, not 0:0!!!!
025D  89 D8                             mov             ax,bx
025F  EB 05                             jmp             L$33
0261                            L$32:
0261  8D 47 02                          lea             ax,0x2[bx] // move near "ptr + 1" to AX
0264  8C D2                             mov             dx,ss // high 16 bits of 32-bit pointer set to SS (=DS) OK
0266                            L$33:
0266  5D                                pop             bp
0267  5F                                pop             di
0268  5E                                pop             si
0269  59                                pop             cx
026A  5B                                pop             bx
026B  CB                                retf
```
Long story short, Open Watcom is doing just what I said to beware of: converting a near pointer to DS:0. So malloc never returns an actual 32-bit 0, it will always return DS:0 when out of memory.

For now, a workaround in your mem.c code would be to inspect just the lower 16-bits of the far pointer returned from either malloc or realloc (but not fmemalloc!). Something like:
```
#define NULLPTR(ptr)  (((unsigned long)(char __far *)(ptr) & 0xFFFF) == 0)
char *p;
...
p = malloc(n); // or realloc, but never fmemalloc
...
if (p == NULL) // BUGGY, can't use
if (p)  // BUGGY, can't use
...
if (NULLPTR(p)) // check for null pointer
if (!NULLPTR(p)) // check for valid non-null pointer
```

File: compiler/system.c (hunk 4 above)
Comment (rafael2k) [reply compiler/system.c:57]: Btw, is there a hard 4096 bytes of maximum memory allocation in fmemalloc?

File: compiler/memmgt.c (hunk 2 above)
Comment (rafael2k) [reply compiler/memmgt.c:37]: It is the first call to fmemalloc(), with 4100 of size, nothing allocated before, it returns NULL, in the very start of x86 execution.

File: compiler/system.c (hunk 4 above)
Comment (ghaerr) [reply compiler/system.c:57]: > is there a hard 4096 bytes of maximum memory allocation in fmemalloc?

No, fmemalloc can allocate as much contiguous memory as may be found in the running system. That's why I find it so strange 4K can't be allocated right after C86 start. The only way this should fail is when the system no longer has 4K contiguous bytes available. With just /bin/init, sh, make and c86 running, I would think there would he a lot more memory available. In order to tell, we would need to run `meminfo`from another login while C86 is running, to inspect memory. I am assuming your ELKS system has 640k main memory?

File: compiler/memmgt.c (hunk 2 above)
Comment (ghaerr) [reply compiler/memmgt.c:37]: > I have a feeling that in the past it used to work (before recent changes in ELKS), but not anymore.

It seemed to me the C86 problem occurred after the fmemalloc was added to it, that it worked fine until you tried to compile chess.c with it and then added the fmemalloc, no?

Comment (rafael2k) [reply compiler/memmgt.c:37]: I added fmemalloc in the very beginning of development. The only difference is that in the beginning I was using fmemalloc for all allocations, and now we have the mem.c logic.

File: compiler/system.c (hunk 4 above)
Comment (rafael2k) [reply compiler/system.c:57]: > > is there a hard 4096 bytes of maximum memory allocation in fmemalloc?
> 
> No, fmemalloc can allocate as much contiguous memory as may be found in the running system. That's why I find it so strange 4K can't be allocated right after C86 start. The only way this should fail is when the system no longer has 4K contiguous bytes available. With just /bin/init, sh, make and c86 running, I would think there would he a lot more memory available. In order to tell, we would need to run `meminfo`from another login while C86 is running, to inspect memory. I am assuming your ELKS system has 640k main memory?

Yes, 640k main memory.

**Comment by rafael2k** [comment]
NULL pointer handling PR:
https://github.com/rafael2k/8086-toolchain/pull/20

File: compiler/memmgt.c (hunk 2 above)
Comment (ghaerr) [reply compiler/memmgt.c:37]: I see, good to know. I think I'll add a capability in ELKS to be able to monitor malloc and fmemalloc requests and sizes, this should give us a better idea as to what is going on, and also allow for easier tuning. The good news is you have everything working now :)
